File: tests/test_analyzer.py
import unittest
from pathlib import Path

from flang_modernizer import ModernizationAnalyzer


ROOT = Path(__file__).resolve().parents[1]


class AnalyzerTests(unittest.TestCase):
    def test_detects_required_patterns(self):
        analysis = ModernizationAnalyzer().analyze_path(ROOT / "examples" / "legacy")
        patterns = {finding.pattern for finding in analysis.findings}
        self.assertIn("arithmetic-if", patterns)
        self.assertIn("computed-goto", patterns)
        self.assertIn("equivalence", patterns)
        self.assertIn("common-block", patterns)
        self.assertIn("implicit-typing", patterns)
        self.assertIn("statement-function", patterns)
        self.assertIn("fixed-form", patterns)
        self.assertIn("assumed-size-array", patterns)
        self.assertIn("entry", patterns)

    def test_common_block_impact_spans_multiple_files(self):
        analysis = ModernizationAnalyzer().analyze_path(ROOT / "examples" / "case_study")
        common_findings = [f for f in analysis.findings if f.pattern == "common-block"]
        self.assertTrue(common_findings)
        self.assertTrue(any(len(f.affected_files) == 2 for f in common_findings))

    def test_risky_constructs_rank_high(self):
        analysis = ModernizationAnalyzer().analyze_path(ROOT / "examples" / "case_study")
        top = analysis.sorted_findings()[0]
        self.assertIn(top.pattern, {"common-block", "equivalence"})


if __name__ == "__main__":
    unittest.main()

File: tests/test_analyzer.cpp
#include "Analyzer.hpp"

#include <cstdlib>
#include <filesystem>
#include <iostream>
#include <set>
#include <string>

namespace {
void require(bool condition, const std::string &message) {
  if (!condition) {
    std::cerr << "FAIL: " << message << "\n";
    std::exit(1);
  }
}

std::set<std::string> patterns(const ProjectAnalysis &analysis) {
  std::set<std::string> result;
  for (const auto &finding : analysis.findings) {
    result.insert(finding.pattern);
  }
  return result;
}
} // namespace

int main() {
  auto root = std::filesystem::current_path();
  ModernizationAnalyzer analyzer;

  auto legacy = analyzer.analyzePath(root / "examples" / "legacy");
  auto found = patterns(legacy);
  require(found.count("arithmetic-if") == 1, "detect arithmetic IF");
  require(found.count("computed-goto") == 1, "detect computed GOTO");
  require(found.count("equivalence") == 1, "detect EQUIVALENCE");
  require(found.count("common-block") == 1, "detect COMMON");
  require(found.count("implicit-typing") == 1, "detect implicit typing");
  require(found.count("statement-function") == 1, "detect statement function");
  require(found.count("fixed-form") == 1, "detect fixed form");
  require(found.count("assumed-size-array") == 1, "detect assumed-size array");
  require(found.count("entry") == 1, "detect ENTRY");


  auto falsePositives = analyzer.analyzePath(root / "examples" / "legacy" / "false_positives.f90");
  auto falsePatterns = patterns(falsePositives);
  require(falsePatterns.count("arithmetic-if") == 0, "ordinary IF THEN is not arithmetic IF");
  require(falsePatterns.count("computed-goto") == 0, "ordinary GOTO is not computed GOTO");
  require(falsePatterns.count("statement-function") == 0, "array assignment is not statement function");

  auto caseStudy = analyzer.analyzePath(root / "examples" / "case_study");
  bool multiFileCommon = false;
  bool riskyTop = false;
  int topPriority = -1;
  std::string topPattern;
  for (const auto &finding : caseStudy.findings) {
    if (finding.pattern == "common-block" && finding.affectedFiles.size() == 2) {
      multiFileCommon = true;
    }
    if (finding.priority > topPriority) {
      topPriority = finding.priority;
      topPattern = finding.pattern;
    }
  }
  riskyTop = topPattern == "common-block" || topPattern == "equivalence";
  require(multiFileCommon, "COMMON impact spans multiple files");
  require(riskyTop, "risky storage constructs rank highest");

  std::cout << "All C++ analyzer tests passed.\n";
  return 0;
}
